[Shape Detection] [Cocoa] Implement first draft of the shape detection API using the Vision framework

litherum opened the pull request:
#### ce99b4d5dd33bd28ffb58899e356841cb5ed7285
<pre>
[Shape Detection] [Cocoa] Implement first draft of the shape detection API using the Vision framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=255803">https://bugs.webkit.org/show_bug.cgi?id=255803</a>
rdar://108380805

Reviewed by Mike Wyrzykowski.

This is the first draft implementation of the shape detection API. This patch makes WebCore.framework link
with Vision.framework. The shape detection API is very very similar to Vision&apos;s API, so the implementation
is almost trivial: each call to detect() just turns into the creation of a VNRequest and a VNRequestHandler,
and calls performRequests() on the request handler.

VNRequest has a callback-based codepath, where you can say -[VNRequest initWithCompletionHandler:], but this
isn&apos;t really suitable for our purposes, at least right out-of-the-box, for 2 reasons:
1. Even if you set up the VNRequest this way, the call to -[VNImageRequestHandler performRequest:error:] is
       still synchronous. It blocks until all the requests are complete, and AFAICT there&apos;s no way to make
       it not block. So, using the callback doesn&apos;t actually free up the main thread to do other work. (The
       purpose of the completion handler is when you pass many requests to a single performRequests: call,
       you can be notified as the requests complete, one by one - *not* so that the call to performRequests:
       can be asynchronous. If you want the call to performRequests: to be asynchronous, you&apos;re supposed to
       call it from a background thread yourself.)
2. The callbacks are called on another thread internal to Vision, so if we wanted to use them, we&apos;d have to
       immediately bounce back to the RemoteRenderingBackend&apos;s worker thread in WebKit2, or the main thread
       in WebKitLegacy. This means the Shape Detection objects would have to have more plumbing to take a
       generic handler which can schedule work on the right thread.
At some point, it probably would make sense to do this plumbing work + calling Vision on a background
thread, but it isn&apos;t really necessary for this first-draft implementation. I haven&apos;t profiled these calls
yet to know if they&apos;re even expensive enough to be worth calling on a background thread or not.

This patch also sets the usesCPUOnly flag, because otherwise the API wouldn&apos;t work in WebKit 2, because the
GPU Process&apos;s sandbox is blocking the connection to aned on Apple Silicon devices (I haven&apos;t tried on Intel
devices, maybe it works using the GPU there?). I&apos;m working with Per Arne to see what we can do about this.
If you manually disable the sandbox, this flag isn&apos;t necessary.

The tests for this patch are at
<a href="https://github.com/web-platform-tests/wpt/commit/b5a2b1e1b0bd51817dca49b49eb3e8825833b216.">https://github.com/web-platform-tests/wpt/commit/b5a2b1e1b0bd51817dca49b49eb3e8825833b216.</a> I will import
them in a follow-up patch.

This patch marks the shape detection as &quot;testable.&quot;

* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm:
(WebCore::ShapeDetection::convertSymbology):
(WebCore::ShapeDetection::convertBarcodeFormat):
(WebCore::ShapeDetection::convertRequestedBarcodeFormatSet):
(WebCore::ShapeDetection::BarcodeDetectorImpl::BarcodeDetectorImpl):
(WebCore::ShapeDetection::request):
(WebCore::ShapeDetection::BarcodeDetectorImpl::getSupportedFormats):
(WebCore::ShapeDetection::BarcodeDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.mm:
(WebCore::ShapeDetection::FaceDetectorImpl::FaceDetectorImpl):
(WebCore::ShapeDetection::convertLandmark):
(WebCore::ShapeDetection::convertLandmarks):
(WebCore::ShapeDetection::FaceDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.mm:
(WebCore::ShapeDetection::TextDetectorImpl::detect):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/263397@main">https://commits.webkit.org/263397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c96e887e8cce90174df08b7955dec0a88dacd4af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5909 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4514 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4856 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4656 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5911 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3965 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/8059 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3682 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4034 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5569 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4210 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4436 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3612 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4528 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3961 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3964 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1125 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1100 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8011 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4633 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4315 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1232 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm:
```diff
@@ -26,25 +26,202 @@
+#include "BarcodeDetectorOptionsInterface.h"
 #include "BarcodeFormatInterface.h"
 #include "DetectedBarcodeInterface.h"
+#include "ImageBuffer.h"
```
These should use `#import`

mwyrzykowski commented on Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm:
```diff
@@ -23,28 +23,205 @@
-BarcodeDetectorImpl::BarcodeDetectorImpl(const BarcodeDetectorOptions&)
+static BarcodeFormat convertSymbology(VNBarcodeSymbology symbology)
+{
+    if (symbology == VNBarcodeSymbologyAztec)
```
Are these guaranteed to be tagged pointers? Otherwise we should use `isEqualToString` or `isEqual` I think.

But it would be preferable to use direct comparison (operator==) as you have it, it avoids many string comparisons, assuming this is guaranteed by the framework.

litherum replied:
That's a good point! I'll ask the Vision team.

If comparing by pointers is correct, then this could be turned into a switch statement anyway.

litherum replied:
Vision team says “use -isEqual”

mwyrzykowski commented on Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm:
```diff
@@ -23,28 +23,205 @@
+        request.get().symbologies = requestedSymbologies.allObjects;
+    }
+
+    auto imageRequestHandler = adoptNS([[VNImageRequestHandler alloc] initWithCGImage:platformImage.get() options:@{ }]);
```
why do we work with CGImages when the framework supports pixel buffers?
```
- (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer options:(NSDictionary<VNImageOption, id> *)options;
- (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer orientation:(CGImagePropertyOrientation)orientation options:(NSDictionary<VNImageOption, id> *)options;
-  (instancetype)initWithCVPixelBuffer:(CVPixelBufferRef)pixelBuffer
                            depthData:(AVDepthData*)depthData
                          orientation:(CGImagePropertyOrientation)orientation
                              options:(NSDictionary<VNImageOption, id> *)options;
```

The data comes from a camera I presume, so we could pass the CVPixelBuffer from the camera stream directly to the vision framework

litherum replied:
Yep! This is a good idea, but I think it's beyond the scope of this initial implementation.

This initial implementation keeps things simple by just having all the various forms drawn into an ImageBuffer, and then we do the recognition on the ImageBuffer. This could be improved; we could skip the copy and just run the recognition directly on the source, but that requires a lot more code, because there are 8 or so different forms the input could be in. So I think we should do that, but not in this patch.

mwyrzykowski replied:
That sounds reasonable to me

litherum committed:
[Shape Detection] [Cocoa] Implement first draft of the shape detection API using the Vision framework
https://bugs.webkit.org/show_bug.cgi?id=255803
rdar://108380805

Reviewed by Mike Wyrzykowski.

This is the first draft implementation of the shape detection API. This patch makes WebCore.framework link
with Vision.framework. The shape detection API is very very similar to Vision's API, so the implementation
is almost trivial: each call to detect() just turns into the creation of a VNRequest and a VNRequestHandler,
and calls performRequests() on the request handler.

VNRequest has a callback-based codepath, where you can say -[VNRequest initWithCompletionHandler:], but this
isn't really suitable for our purposes, at least right out-of-the-box, for 2 reasons:
1. Even if you set up the VNRequest this way, the call to -[VNImageRequestHandler performRequest:error:] is
       still synchronous. It blocks until all the requests are complete, and AFAICT there's no way to make
       it not block. So, using the callback doesn't actually free up the main thread to do other work. (The
       purpose of the completion handler is when you pass many requests to a single performRequests: call,
       you can be notified as the requests complete, one by one - *not* so that the call to performRequests:
       can be asynchronous. If you want the call to performRequests: to be asynchronous, you're supposed to
       call it from a background thread yourself.)
2. The callbacks are called on another thread internal to Vision, so if we wanted to use them, we'd have to
       immediately bounce back to the RemoteRenderingBackend's worker thread in WebKit2, or the main thread
       in WebKitLegacy. This means the Shape Detection objects would have to have more plumbing to take a
       generic handler which can schedule work on the right thread.
At some point, it probably would make sense to do this plumbing work + calling Vision on a background
thread, but it isn't really necessary for this first-draft implementation. I haven't profiled these calls
yet to know if they're even expensive enough to be worth calling on a background thread or not.

This patch also sets the usesCPUOnly flag, because otherwise the API wouldn't work in WebKit 2, because the
GPU Process's sandbox is blocking the connection to aned on Apple Silicon devices (I haven't tried on Intel
devices, maybe it works using the GPU there?). I'm working with Per Arne to see what we can do about this.
If you manually disable the sandbox, this flag isn't necessary.

The tests for this patch are at
https://github.com/web-platform-tests/wpt/commit/b5a2b1e1b0bd51817dca49b49eb3e8825833b216. I will import
them in a follow-up patch.

This patch marks the shape detection as "testable."

* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm:
(WebCore::ShapeDetection::convertSymbology):
(WebCore::ShapeDetection::convertBarcodeFormat):
(WebCore::ShapeDetection::convertRequestedBarcodeFormatSet):
(WebCore::ShapeDetection::BarcodeDetectorImpl::BarcodeDetectorImpl):
(WebCore::ShapeDetection::request):
(WebCore::ShapeDetection::BarcodeDetectorImpl::getSupportedFormats):
(WebCore::ShapeDetection::BarcodeDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.mm:
(WebCore::ShapeDetection::FaceDetectorImpl::FaceDetectorImpl):
(WebCore::ShapeDetection::convertLandmark):
(WebCore::ShapeDetection::convertLandmarks):
(WebCore::ShapeDetection::FaceDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.mm:
(WebCore::ShapeDetection::TextDetectorImpl::detect):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/263397@main
